system/doc/tutorial/debugging.xml

```diff
@@ -0,0 +1,345 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!DOCTYPE chapter SYSTEM "chapter.dtd">
+<chapter>
+  <header>
+    <copyright>
+      <year>2022</year>
+      <holder>Ericsson AB. All Rights Reserved.</holder>
+    </copyright>
+    <legalnotice>
+      Licensed under the Apache License, Version 2.0 (the "License");
+      you may not use this file except in compliance with the License.
+      You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+      Unless required by applicable law or agreed to in writing, software
+      distributed under the License is distributed on an "AS IS" BASIS,
+      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+      See the License for the specific language governing permissions and
+      limitations under the License.
+
+    </legalnotice>
+
+    <title>Debugging NIFs and Port Drivers</title>
+    <prepared/>
+    <docno/>
+    <date/>
+    <rev/>
+    <file>debugging.xml</file>
+  </header>
+
+  <section>
+    <title>With great power comes great responsibilty</title>
+    <p>
+      NIFs and port driver code run inside the Erlang VM OS process (the
+      "Beam"). To maximize performance the code is called directly by the same
+      threads executing Erlang beam code and has full access to all the memory
+      of the OS process. A buggy NIF/driver can thus make severe damage by
+      corrupting memory.
+    </p>
+    <p>
+      In a best case scenario such memory corruption is detected immediately
+      causing the Beam to crash generating a core dump file which can be
+      analyzed to find the bug. However, it is very common for memory corruption
+      bugs to not be immediately detected when the faulty write happens, but
+      instead much later, for example when the calling Erlang process is garbage
+      collected. When that happens it can be very hard to find the root cause of
+      the memory corruption by analysing the core dump. All traces that could
+      have indicated which specific buggy NIF/driver that caused the corruption
+      may be long gone.
+    </p>
+  </section>
+  <section>
+    <title>The debug emulator</title>
+    <p>
+      One way to make debugging easier is to run an emulator built with target
+      <c>debug</c>. It will
+    </p>
+    <list type="bulleted">
+      <item>
+	<p>
+	  <em>Increase probability of detecting bugs earlier</em>. It contains a
+	  lot more runtime checks to ensure correct use of internal interfaces
+	  and data structures.
+	</p>
+      </item>
+      <item>
+	<p>
+	  <em>Generate a core dump that is easier to analyze</em>. Compiler
+	  optimizations are turned off, which can otherwise make local variables
+	  to be "optmized away" and therefor much harder or even impossible to
+	  see.
+	</p>
```

Comment:
I understand what you mean, but the word order is strange (and with some typos), what about:

Compiler optimizations are turned off, which stops the compiler from "optimizing away" variables, thus making it easier to inspect the state of the NIF.

Reply:
Thanks.